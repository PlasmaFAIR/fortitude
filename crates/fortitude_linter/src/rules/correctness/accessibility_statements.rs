use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for missing `private` or `public` accessibility statements in modules
///
/// ## Why is this bad?
/// The `private` statement makes all entities (variables, types, procedures)
/// private by default, requiring an explicit `public` attribute to make them
/// available. As well as improving encapsulation between modules, this also
/// makes it possible to detect unused entities.
///
/// A `public` statement in a module does not change the default behaviour,
/// and therefore all entities will be available from outside the module
/// unless they are individually given a `private` attribute. This brings
/// all of the same downsides as the default behaviour, but an explicit
/// `public` statement makes it clear that the programmer is choosing
/// this behaviour intentionally.  
#[derive(ViolationMetadata)]
pub(crate) struct MissingAccessibilityStatement {
    name: String,
}

impl Violation for MissingAccessibilityStatement {
    #[derive_message_formats]
    fn message(&self) -> String {
        format!(
            "module '{}' missing default accessibility statement",
            self.name
        )
    }
}

impl AstRule for MissingAccessibilityStatement {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let module = node.parent()?;

        let bare_private_statement = match module.child_with_name("private_statement") {
            Some(statement) => statement.named_child(0).is_none(),
            None => false,
        };

        let bare_public_statement = match module.child_with_name("public_statement") {
            Some(statement) => statement.named_child(0).is_none(),
            None => false,
        };

        // No statement whatsoever
        if !bare_private_statement && !bare_public_statement {
            let name = node.named_child(0)?.to_text(src.source_text())?.to_string();
            return some_vec![Diagnostic::from_node(
                MissingAccessibilityStatement { name },
                node
            )];
        }

        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["module_statement"]
    }
}

/// ## What it does
/// Checks if the default accessibility in modules is set to `public`
///
/// ## Why is this bad?
/// The `public` statement makes all entities (variables, types, procedures)
/// public by default. This decreases encapsulation and makes it more likely to
/// accidentally expose more than necessary. Public accessibility also makes
/// it harder to detect unused entities, which can often be indicative of
/// errors within the code.
#[derive(ViolationMetadata)]
pub(crate) struct DefaultPublicAccessibility {
    name: String,
}

impl Violation for DefaultPublicAccessibility {
    #[derive_message_formats]
    fn message(&self) -> String {
        format!("module '{}' has default `public` accessibility", self.name)
    }
}

impl AstRule for DefaultPublicAccessibility {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // Bare `public` statement`
        if node.named_child(0).is_none() {
            let module = node.parent()?;
            let statement = module.child_with_name("module_statement")?;
            let name = statement
                .child_with_name("name")?
                .to_text(src.source_text())?
                .to_string();
            return some_vec![Diagnostic::from_node(
                DefaultPublicAccessibility { name },
                node
            )];
        }

        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["public_statement"]
    }
}
