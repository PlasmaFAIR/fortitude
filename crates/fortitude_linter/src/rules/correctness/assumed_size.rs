use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use itertools::Itertools;
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What does it do?
/// Checks for assumed size variables
///
/// ## Why is this bad?
/// Assumed size dummy arguments declared with a star `*` as the size should be
/// avoided. There are several downsides to assumed size, the main one being
/// that the compiler is not able to determine the array bounds, so it is not
/// possible to check for array overruns or to use the array in whole-array
/// expressions.
///
/// Instead, prefer assumed shape arguments, as the compiler is able to keep track of
/// the upper bounds automatically, and pass this information under the hood. It also
/// allows use of whole-array expressions, such as `a = b + c`, where `a, b, c` are
/// all arrays of the same shape.
///
/// Instead of:
///
/// ```f90
/// subroutine process_array(array)
///     integer, dimension(*), intent(in) :: array
///     ...
/// ```
///
/// use:
///
/// ```f90
/// subroutine process_array(array)
///     integer, dimension(:), intent(in) :: array
///     ...
/// ```
///
/// Note that this doesn't apply to `character` types, where `character(len=*)` is
/// actually the most appropriate specification for `intent(in)` arguments! This is
/// because `character(len=:)` must be either a `pointer` or `allocatable`.
#[derive(ViolationMetadata)]
pub(crate) struct AssumedSize {
    name: String,
}

impl Violation for AssumedSize {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("'{name}' has assumed size")
    }
}
impl AstRule for AssumedSize {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        let declaration = node
            .ancestors()
            .find(|parent| parent.kind() == "variable_declaration")?;

        // Deal with `character([len=]*)` elsewhere
        if let Some(dtype) = declaration.parse_intrinsic_type() {
            let is_character = dtype.to_lowercase() == "character";
            let is_kind = node.ancestors().any(|parent| parent.kind() == "kind");
            if is_character && is_kind {
                return None;
            }
        }

        // Assumed size ok for parameters
        if declaration
            .children_by_field_name("attribute", &mut declaration.walk())
            .filter_map(|attr| attr.to_text(src))
            .any(|attr_name| attr_name.to_lowercase() == "parameter")
        {
            return None;
        }

        // Are we looking at something declared like `array(*)`?
        if let Some(sized_decl) = node
            .ancestors()
            .find(|parent| parent.kind() == "sized_declarator")
        {
            let identifier = sized_decl.child_with_name("identifier")?;
            let name = identifier.to_text(src)?.to_string();
            return some_vec![Diagnostic::from_node(Self { name }, node)];
        }

        // Collect things that look like `dimension(*)` -- this
        // applies to all identifiers on this line
        let all_decls = declaration
            .children_by_field_name("declarator", &mut declaration.walk())
            .filter_map(|declarator| {
                let identifier = match declarator.kind() {
                    "identifier" => Some(declarator),
                    "sized_declarator" => declarator.child_with_name("identifier"),
                    _ => None,
                }?;
                identifier.to_text(src)
            })
            .map(|name| name.to_string())
            .map(|name| Diagnostic::from_node(Self { name }, node))
            .collect_vec();

        Some(all_decls)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["assumed_size"]
    }
}

/// ## What does it do?
/// Checks `character` dummy arguments have `intent(in)` only
///
/// ## Why is this bad?
/// Character dummy arguments with an assumed size should only have `intent(in)`, as
/// this can cause data loss with `intent([in]out)`. For example:
///
/// ```f90
/// program example
///   character(len=3) :: short_text
///   call set_text(short_text)
///   print*, short_text
/// contains
///   subroutine set_text(text)
///     character(*), intent(out) :: text
///     text = "longer than 3 characters"
///   end subroutine set_text
/// end program
/// ```
///
/// Here, `short_text` will only contain the truncated "lon".
///
/// To handle dynamically setting `character` sizes, use `allocatable` instead:
///
/// ```f90
/// program example
///   character(len=:), allocatable :: allocatable_text
///   call set_text(allocatable_text)
///   print*, allocatable_text
/// contains
///   subroutine set_text(text)
///     character(len=:), allocatable, intent(out) :: text
///     text = "longer than 3 characters"
///   end subroutine set_text
/// end program
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct AssumedSizeCharacterIntent {
    name: String,
}

impl Violation for AssumedSizeCharacterIntent {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("character '{name}' has assumed size but does not have `intent(in)`")
    }
}
impl AstRule for AssumedSizeCharacterIntent {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // TODO: This warning will also catch:
        // - non-dummy arguments -- these are always invalid, should be a separate warning?

        let declaration = node
            .ancestors()
            .find(|parent| parent.kind() == "variable_declaration")?;

        // Only applies to `character`
        if declaration.parse_intrinsic_type()?.to_lowercase() != "character" {
            return None;
        }

        // Handle `character*(*)` elsewhere -- note this just skips emitting a warning
        // for the first `*`, we'll still get one for the second `*`, but this is desired
        if let Some(sibling) = node.next_named_sibling() {
            if sibling.kind() == "assumed_size" {
                return None;
            }
        }

        let attrs_as_text = declaration
            .children_by_field_name("attribute", &mut declaration.walk())
            .filter_map(|attr| attr.to_text(src))
            .map(|attr| attr.to_lowercase())
            .collect_vec();

        // Assumed size ok for parameters
        if attrs_as_text.iter().any(|attr| attr == "parameter") {
            return None;
        }

        // Ok for `intent(in)` only
        if let Some(intent) = attrs_as_text.iter().find(|attr| attr.starts_with("intent")) {
            let intent = intent.split_whitespace().collect_vec().join("");
            if intent == "intent(in)" {
                return None;
            }
        }

        // Collect all declarations on this line
        let all_decls = declaration
            .children_by_field_name("declarator", &mut declaration.walk())
            .filter_map(|declarator| {
                let identifier = match declarator.kind() {
                    "identifier" => Some(declarator),
                    "sized_declarator" => declarator.child_with_name("identifier"),
                    _ => None,
                }?;
                identifier.to_text(src)
            })
            .map(|name| name.to_string())
            .map(|name| Diagnostic::from_node(Self { name }, node))
            .collect_vec();

        Some(all_decls)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["assumed_size"]
    }
}
