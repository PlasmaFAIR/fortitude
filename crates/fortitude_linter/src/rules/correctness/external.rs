use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

use crate::{ast::FortitudeNode, settings::Settings, AstRule, FromAstNode};

/// ## What does it do?
/// Checks for procedures declared with just `external`
///
/// ## Why is this bad?
/// Compilers are unable to check external procedures without an explicit
/// interface for errors such as wrong number or type of arguments.
///
/// If the procedure is in your project, put it in a module (see
/// `external-function`), or write an explicit interface.
#[derive(ViolationMetadata)]
pub(crate) struct ExternalProcedure {
    name: String,
}

impl Violation for ExternalProcedure {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("'{name}' declared as `external`")
    }

    fn fix_title(&self) -> Option<String> {
        Some(format!("Write an explicit interface"))
    }
}

impl AstRule for ExternalProcedure {
    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        if node
            .child_with_name("type_qualifier")?
            .to_text(source.source_text())?
            .to_lowercase()
            != "external"
        {
            return None;
        }

        let name = node
            .child_by_field_name("declarator")?
            .to_text(source.source_text())?
            .to_string();
        some_vec!(Diagnostic::from_node(Self { name }, node))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["variable_modification"]
    }
}

/// ## What it does
/// Checks for any functions and subroutines not defined within modules (or one
/// of a few acceptable alternatives).
///
/// ## Why is this bad?
/// Functions and subroutines should be contained within (sub)modules or programs.
/// Fortran compilers are unable to perform type checks and conversions on functions
/// defined outside of these scopes, and this is a common source of bugs.
#[derive(ViolationMetadata)]
pub(crate) struct ProcedureNotInModule {
    procedure: String,
}

impl Violation for ProcedureNotInModule {
    #[derive_message_formats]
    fn message(&self) -> String {
        let ProcedureNotInModule { procedure } = self;
        format!("{procedure} not contained within (sub)module or program")
    }
}

impl AstRule for ProcedureNotInModule {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if node.parent()?.kind() == "translation_unit" {
            let procedure_stmt = node.child(0)?;
            let procedure = node.kind().to_string();
            return some_vec![Diagnostic::from_node(
                ProcedureNotInModule { procedure },
                &procedure_stmt
            )];
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["function", "subroutine"]
    }
}
