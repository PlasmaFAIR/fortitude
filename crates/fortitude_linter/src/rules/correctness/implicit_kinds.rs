use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for `real` variables that don't have their kind explicitly specified.
///
/// ## Why is this bad?
/// Real variable declarations without an explicit kind will have a compiler/platform
/// dependent precision, which hurts portability and may lead to surprising loss of
/// precision in some cases. Although the default `real` will map to a 32-bit floating
/// point number on most systems, this is not guaranteed.
///
/// It is recommended to always be explicit about the precision required by `real`
/// variables. This can be done by setting their 'kinds' using integer parameters
/// chosen in one of the following ways:
///
/// ```f90
/// ! Set using iso_fortran_env
/// use, intrinsic :: iso_fortran_env, only: sp => real32, dp => real64
/// ! Using selected_real_kind
/// integer, parameter :: sp = selected_real_kind(6, 37)
/// integer, parameter :: dp = selected_real_kind(15, 307)
/// ! For C-compatibility:
/// use, intrinsic :: iso_c_binding, only: sp => c_float, dp => c_double
///
/// ! Declaring real variables:
/// real(sp) :: single
/// real(dp) :: double
/// ```
///
/// It is also common for Fortran developers to set a 'working precision' `wp`,
/// which is set to either `sp` or `dp` and used throughout a project. This can
/// then be easily toggled depending on the user's needs.
///
/// ## References
/// - [Fortran-Lang Best Practices on Floating Point Numbers](https://fortran-lang.org/learn/best_practices/floating_point/)
#[derive(ViolationMetadata)]
pub(crate) struct ImplicitRealKind {
    dtype: String,
}

impl Violation for ImplicitRealKind {
    #[derive_message_formats]
    fn message(&self) -> String {
        let ImplicitRealKind { dtype } = self;
        format!("{dtype} has implicit kind")
    }
}

impl AstRule for ImplicitRealKind {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let dtype = node.child(0)?.to_text(src.source_text())?.to_lowercase();

        if !matches!(dtype.as_str(), "real" | "complex") {
            return None;
        }

        if node.child_by_field_name("kind").is_some() {
            return None;
        }

        if node.parent()?.kind() == "derived_type" && node.next_named_sibling()?.kind() == "kind" {
            return None;
        }

        some_vec![Diagnostic::from_node(ImplicitRealKind { dtype }, node)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["intrinsic_type"]
    }
}
