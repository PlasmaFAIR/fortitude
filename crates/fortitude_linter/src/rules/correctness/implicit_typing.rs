/// Defines rules that raise errors if implicit typing is in use.
use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{ViolationMetadata, derive_message_formats};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

pub fn implicit_statement_is_none(node: &Node) -> bool {
    if let Some(child) = node.child(1) {
        return child.kind() == "none";
    }
    false
}

pub fn has_implicit_none(node: &Node) -> bool {
    if let Some(child) = node.child_with_name("implicit_statement") {
        return implicit_statement_is_none(&child);
    }
    false
}

fn insert_implicit_none(node: &Node, src: &SourceFile) -> Option<Fix> {
    // Find suitable place to insert `implicit none`, the line
    // after the last `use` statement, if any
    let last_use_statement_range = node
        .named_children(&mut node.walk())
        .filter_map(|child| {
            if child.kind() == "use_statement" {
                Some(child.textrange())
            } else {
                None
            }
        })
        .last()
        .or(Some(node.child(0)?.textrange()))?;

    // Get the start and end of the line
    let source_code = src.to_source_code();
    let source_location = source_code.source_location(last_use_statement_range.start());
    let line_start = source_code.line_start(source_location.row);
    let line_end = source_code.line_end(source_location.row);

    // TODO(peter): determine indentation of file using `Stylist` struct
    let indent = (last_use_statement_range.start() - line_start).to_usize();
    let insert = format!("{:indent$}implicit none\n", "");

    let edit = Edit::insertion(insert, line_end);

    Some(Fix::unsafe_edit(edit))
}

/// ## What does it do?
/// Checks for missing `implicit none`.
///
/// ## Why is this bad?
/// Very early Fortran determined the type of variables implicitly
/// from the first character of their name which saved lines in the
/// days of punchcards, and for backwards compatibility this is still
/// the default behaviour. However, the major downside is that typos
/// can silently introduce undefined variables and lead to hard to
/// track down bugs. For example:
///
/// ```f90
/// do i = 1, 10
///     print*, in
/// end do
/// ```
///
/// will print garbage.
///
/// 'implicit none' should be used in all modules and programs, as
/// implicit typing reduces the readability of code and increases the
/// chances of typing errors.
#[derive(ViolationMetadata)]
pub(crate) struct ImplicitTyping {
    entity: String,
}

impl Violation for ImplicitTyping {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { entity } = self;
        format!("{entity} missing 'implicit none'")
    }

    fn fix_title(&self) -> Option<String> {
        Some("Insert `implicit none`".to_string())
    }
}
impl AstRule for ImplicitTyping {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if has_implicit_none(node) {
            return None;
        }
        let entity = node.kind().to_string();
        let block_stmt = node.child(0)?;

        some_vec![
            Diagnostic::from_node(Self { entity }, &block_stmt)
                .with_fix(insert_implicit_none(node, src)?)
        ]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["module", "submodule", "program"]
    }
}

/// ## What it does
/// Checks for missing `implicit none` in interfaces
///
/// ## Why is this bad?
/// Interface functions and subroutines require 'implicit none', even if they are
/// inside a module that uses 'implicit none'.
#[derive(ViolationMetadata)]
pub(crate) struct InterfaceImplicitTyping {
    name: String,
}

impl Violation for InterfaceImplicitTyping {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("interface '{name}' missing 'implicit none'")
    }

    fn fix_title(&self) -> Option<String> {
        Some("Insert `implicit none`".to_string())
    }
}

impl AstRule for InterfaceImplicitTyping {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let parent = node.parent()?;
        if parent.kind() == "interface" && !has_implicit_none(node) {
            let name = node.kind().to_string();
            let interface_stmt = node.child(0)?;
            return some_vec![
                Diagnostic::from_node(Self { name }, &interface_stmt)
                    .with_fix(insert_implicit_none(node, src)?)
            ];
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["function", "subroutine"]
    }
}

/// ## What it does
/// Checks if `implicit none` is missing `external`
///
/// ## Why is this bad?
/// `implicit none` disables implicit types of variables but still allows
/// implicit interfaces for procedures. Fortran 2018 added the ability to also
/// forbid implicit interfaces through `implicit none (external)`, enabling the
/// compiler to check the number and type of arguments and return values.
///
/// `implicit none` is equivalent to `implicit none (type)`, so the full
/// statement should be `implicit none (type, external)`.
#[derive(ViolationMetadata)]
pub(crate) struct ImplicitExternalProcedures {}

impl Violation for ImplicitExternalProcedures {
    #[derive_message_formats]
    fn message(&self) -> String {
        "'implicit none' missing 'external'".to_string()
    }

    fn fix_title(&self) -> Option<String> {
        Some("Add `(external)` to 'implicit none'".to_string())
    }
}

impl AstRule for ImplicitExternalProcedures {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if !implicit_statement_is_none(node) {
            return None;
        }

        let text = node.to_text(src.source_text())?.to_lowercase();

        if !text.contains("external") {
            let edit = if let Some(type_node) = node
                .children(&mut node.walk())
                .find(|child| child.to_text(src.source_text()).unwrap().to_lowercase() == "type")
            {
                // Seems unlikely someone would have `implicit none (type)`
                // without `external` -- is that a sign they _explicitly_ don't
                // want it? That's probably still unwise though
                Edit::insertion(", external".to_string(), type_node.end_textsize())
            } else {
                Edit::insertion(" (type, external)".to_string(), node.end_textsize())
            };
            let fix = Fix::unsafe_edit(edit);

            some_vec!(Diagnostic::from_node(Self {}, node).with_fix(fix))
        } else {
            None
        }
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["implicit_statement"]
    }
}
