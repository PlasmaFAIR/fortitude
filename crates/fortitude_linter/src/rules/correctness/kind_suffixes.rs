use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use lazy_regex::regex_is_match;
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for floating point literal constants that don't have their kinds
/// explicitly specified.
///
/// ## Why is this bad?
/// Floating point literals use the default 'real' kind unless given an explicit
/// kind suffix. This can cause surprising loss of precision:
///
/// ```f90
/// use, intrinsic :: iso_fortran_env, only: dp => real64
///
/// real(dp), parameter :: pi_1 = 3.14159265358979
/// real(dp), parameter :: pi_2 = 3.14159265358979_dp
///
/// print *, pi_1  ! Gives: 3.1415927410125732
/// print *, pi_2  ! Gives: 3.1415926535897900
/// ```
///
/// There are cases where the difference in precision doesn't matter, such
/// as:
///
/// ```f90
/// real(dp) :: x, y
///
/// x = 1.0
/// y = real(2.0, kind=dp)
/// ```
///
/// A case where a missing suffix may be intentional is when using a `kind`
/// statement:
///
/// ```f90
/// integer, parameter :: sp = kind(0.0)
/// ```
///
/// This rule will try to avoid catching these case. However, even for 'nice'
/// numbers, it's possible to accidentally lose precision in surprising ways:
///
/// ```f90
/// real(dp) :: x
///
/// x = sqrt(2.0)
/// ```
///
/// This rule will therefore require an explicit kind statement in the majority
/// of cases where a floating point literal is found in an expression.
///
/// ## References
/// - [Fortran-Lang Best Practices on Floating Point Numbers](https://fortran-lang.org/learn/best_practices/floating_point/)
#[derive(ViolationMetadata)]
pub(crate) struct NoRealSuffix {
    literal: String,
}

impl Violation for NoRealSuffix {
    #[derive_message_formats]
    fn message(&self) -> String {
        let NoRealSuffix { literal } = self;
        format!("real literal {literal} missing kind suffix")
    }
}

impl AstRule for NoRealSuffix {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // Given a number literal, match anything with one or more of a decimal place or
        // an exponentiation e or E. There should not be an underscore present.
        // Exponentiation with d or D are ignored, and should be handled with the
        // rule `double_precision_literal`.
        let txt = node.to_text(src.source_text())?;
        if !regex_is_match!(r"^(\d*\.\d*|\d*\.*\d*[eE]-?\d+)$", txt) {
            return None;
        }

        // Determine the immediate context in which we've found the literal.
        let mut parent = node.parent()?;
        while matches!(
            parent.kind(),
            "unary_expression" | "parenthesized_expression" | "complex_literal"
        ) {
            parent = parent.parent()?;
        }
        let grandparent = parent.parent()?;

        // Check for loss of precision
        // FIXME: This precision loss test isn't the most reliable
        let value_64: f64 = txt.parse().ok()?;
        let value_32: f32 = txt.parse().ok()?;
        let no_loss = value_64 == 0.0
            || (((value_32 as f64) - value_64) / value_64).abs() < 2.0 * f64::EPSILON;

        // Ok if being used in a direct assignment, provided no loss of precision
        // can occur.
        if matches!(parent.kind(), "assignment_statement" | "init_declarator") && no_loss {
            return None;
        }

        // Ok if being used in a kind statement or a type cast.
        // In the latter case, warnings should still be raised if precision would be
        // lost.
        // If it's the sole argument in a function call, the first parent must be
        // "argument_list", and the second must be "call_expression".
        if grandparent.kind() == "call_expression" {
            if let Some(identifier) = grandparent.child_with_name("identifier") {
                let name = identifier.to_text(src.source_text())?.to_lowercase();
                if name == "kind"
                    || (no_loss
                        && matches!(name.as_str(), "real" | "cmplx" | "dbl" | "int" | "logical"))
                {
                    return None;
                }
            }
        }

        let literal = txt.to_string();
        some_vec![Diagnostic::from_node(NoRealSuffix { literal }, node)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["number_literal"]
    }
}
