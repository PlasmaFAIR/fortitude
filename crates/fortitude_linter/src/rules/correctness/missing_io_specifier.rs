use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for missing action specifier when opening files.
///
/// ## Why is this bad?
/// By default, files are opened in `readwrite` mode, but this may not be the
/// programmer's intent. Explicitly specifying `read`, `write` or `readwrite`
/// makes it clear how the file is intended to be used, and prevents the
/// accidental overwriting of input data.
#[derive(ViolationMetadata)]
pub(crate) struct MissingActionSpecifier {}

impl Violation for MissingActionSpecifier {
    #[derive_message_formats]
    fn message(&self) -> String {
        "file opened without action specifier".to_string()
    }

    fn fix_title(&self) -> Option<String> {
        Some("Add 'action=read', 'action=write', or 'action=readwrite'".to_string())
    }
}

impl AstRule for MissingActionSpecifier {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if node.kwarg_exists("action", src.source_text()) {
            return None;
        }
        some_vec![Diagnostic::from_node(Self {}, node)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["open_statement"]
    }
}
