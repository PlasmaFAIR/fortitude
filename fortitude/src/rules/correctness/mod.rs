pub mod accessibility_statements;
pub mod assumed_size;
pub mod derived_default_init;
pub mod exit_labels;
pub mod external;
pub mod implicit_kinds;
pub mod implicit_typing;
pub mod init_decls;
pub mod intent;
pub mod kind_suffixes;
pub mod magic_numbers;
pub mod missing_io_specifier;
pub mod select_default;
pub mod trailing_backslash;
pub mod use_statements;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::correctness::exit_labels;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::ImplicitTyping, Path::new("C001.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002.f90"))]
    #[test_case(Rule::ImplicitExternalProcedures, Path::new("C003.f90"))]
    #[test_case(Rule::MissingDefaultCase, Path::new("C011.f90"))]
    #[test_case(Rule::NoRealSuffix, Path::new("C021.f90"))]
    #[test_case(Rule::ImplicitRealKind, Path::new("C022.f90"))]
    #[test_case(Rule::MagicNumberInArraySize, Path::new("C031.f90"))]
    #[test_case(Rule::MagicIoUnit, Path::new("C032.f90"))]
    #[test_case(Rule::MissingActionSpecifier, Path::new("C041.f90"))]
    #[test_case(Rule::TrailingBackslash, Path::new("C051.F90"))]
    #[test_case(Rule::MissingIntent, Path::new("C061.f90"))]
    #[test_case(Rule::AssumedSize, Path::new("C071.f90"))]
    #[test_case(Rule::AssumedSizeCharacterIntent, Path::new("C072.f90"))]
    #[test_case(Rule::InitialisationInDeclaration, Path::new("C081.f90"))]
    #[test_case(Rule::PointerInitialisationInDeclaration, Path::new("C082.f90"))]
    #[test_case(Rule::ExternalProcedure, Path::new("C091.f90"))]
    #[test_case(Rule::ProcedureNotInModule, Path::new("C092.f90"))]
    #[test_case(Rule::MissingDefaultPointerInitalisation, Path::new("C101.f90"))]
    #[test_case(Rule::UseAll, Path::new("C121.f90"))]
    #[test_case(Rule::MissingIntrinsic, Path::new("C122.f90"))]
    #[test_case(Rule::MissingAccessibilityStatement, Path::new("C131.f90"))]
    #[test_case(Rule::DefaultPublicAccessibility, Path::new("C132.f90"))]
    #[test_case(Rule::MissingExitOrCycleLabel, Path::new("C141.f90"))]
    #[test_case(Rule::ExitOrCycleInUnlabelledLoop, Path::new("C142.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::ImplicitTyping, Path::new("C001_ok.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test]
    fn warn_c142_on_nested_loops_only() -> Result<()> {
        let rule_code = Rule::ExitOrCycleInUnlabelledLoop;
        let path = Path::new("C142.f90");
        let snapshot = format!(
            "{}_{}_nested_loops_only",
            rule_code.as_ref(),
            path.to_string_lossy()
        );
        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                exit_unlabelled_loops: exit_labels::settings::Settings {
                    allow_unnested_loops: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
