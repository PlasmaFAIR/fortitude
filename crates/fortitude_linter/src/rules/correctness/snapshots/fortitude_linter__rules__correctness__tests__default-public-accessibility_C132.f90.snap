---
source: fortitude/src/rules/correctness/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/correctness/C132.f90:10:3: C132 module 'bad_module' has default `public` accessibility
   |
 8 | module bad_module
 9 |   ! This is bad
10 |   public
   |   ^^^^^^ C132
11 | end module bad_module
   |
